crates/stark_poseidon/src/hash.rs

```diff
@@ -0,0 +1,75 @@
+use crate::poseidon::{permute, PoseidonState};
+use stark_curve::{FieldElement, FIELD_ZERO};
+
+/// Hashes a number of messages using the Poseidon hash
+pub fn poseidon_hash(msgs: &[FieldElement]) -> FieldElement {
+    let mut state = [FIELD_ZERO, FIELD_ZERO, FIELD_ZERO];
+    for msg in msgs {
+        state[0] += msg;
+        permute(&mut state);
+    }
+    state[0]
+}
+
+/// The PoseidonHasher can build up a hash by appending to state
+pub struct PoseidonHasher {
+    state: PoseidonState,
+}
+
+impl PoseidonHasher {
+    /// Creates a new PoseidonHasher
+    pub fn new() -> PoseidonHasher {
+        PoseidonHasher {
+            state: [FIELD_ZERO, FIELD_ZERO, FIELD_ZERO],
+        }
+    }
+
+    /// Absorbs message into the hash
+    pub fn write(&mut self, msg: FieldElement) {
+        self.state[0] += msg;
+        permute(&mut self.state);
+    }
+
+    /// Extracts a single output
+    pub fn finish(self) -> FieldElement {
+        self.state[0]
+    }
+}
+
+impl Default for PoseidonHasher {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{poseidon_hash, PoseidonHasher};
+    use stark_curve::{FieldElement, FIELD_ONE, FIELD_THREE, FIELD_TWO, FIELD_ZERO};
+    use stark_hash::Felt;
+
+    #[test]
+    fn test_sponge() {
+        let result = FieldElement::from(
+            Felt::from_hex_str("30ACF0EF3C4549E3684B19652015FD9EADF7BDBA2A1A46CB29B9E18E6622296")
+                .unwrap(),
+        );
+
+        // Construct messages, the first few integers
+        let msgs = [FIELD_ZERO, FIELD_ONE, FIELD_TWO, FIELD_THREE];
+
+        // Construct hash from hasher
+        let mut hasher = PoseidonHasher::new();
+        for msg in msgs {
+            hasher.write(msg);
+        }
+        let hasher_result = hasher.finish();
+
+        // Construct hash from hash function
+        let hash_result = poseidon_hash(&msgs);
+
+        // Check they are equal
+        assert_eq!(hasher_result, hash_result);
+        assert_eq!(result, hash_result);
+    }
```

Comment:
Will add some stuff to the readme, but basically a sponge is a function that can absorb material ("input to hash") and squeeze ("output random/mixed stuff"), [see wiki](https://en.wikipedia.org/wiki/Sponge_function).

From a higher perspective, it's a construction used for modern symmetric crypto primitives (hash, encryption etc) such as SHA3, Poseidon etc. Older stuff like SHA2,MD5 etc. typically use the Merkle-Damgård construction. In short, one main difference is it assumes access to a secure permutation rather than a compression function, hence the main thing here is a Poseidon permutation, and then a Sponge around it. See further in next comment.